clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -232,6 +232,15 @@ public RuntimeException lastError() {
         return lastError;
     }
 
+    public synchronized boolean ensurePartitionAdded(TopicPartition tp) {
+        if (isInTransaction() && !partitionsInTransaction.contains(tp)) {
+            transitionToFatalError(new IllegalStateException("Attempted to dequeue a record batch to send " +
```

Comment:
Makes sense. I was following the ordinary rule for using RTE for any unexpected bugs, but this is definitely better in operations.